`<calc-constant>` - New CSS data type (#20856)
mdn/content: files/en-us/web/css/calc-constant/index.md

## Syntax

The `<calc-constant>` type defines numeric constants that can be used in [CSS math functions](/en-US/docs/Web/CSS/CSS_Functions#math_functions).

### Values

- `e`

  - : The base of the natural logarithm, approximately equal to `2.7182818284590452354`.

- `pi`

  - : The ratio of a circle's circumference to its diameter, approximately equal to `3.1415926535897932`.

- `infinity` & `-infinity`

  - : An infinite value, used to indicate the largest/smallest possible value.

- `NaN`

  - : A value representing "Not a Number" canonical casing.

### Formal syntax

{{CSSSyntax}}

## Description

Mathematical constants can only be used inside [CSS math functions](/en-US/docs/Web/CSS/CSS_Functions#math_functions) for calculations. Math constants are not CSS keywords, but if they are used outside of a calculation, they're treated like any other keyword.

For example:

- `animation-name: pi;` refers to an animation named "pi", not the `pi` numeric constant.
- `line-height: e;` is invalid, but `line-height: calc(e);` is valid.

> 
> 

Reviewer: ```suggestion

```

## Syntax

The `<calc-constant>` type defines numeric constants that can be used in [CSS math functions](/en-US/docs/Web/CSS/CSS_Functions#math_functions).

### Values

- `e`

  - : The base of the natural logarithm, approximately equal to `2.7182818284590452354`.

- `pi`

  - : The ratio of a circle's circumference to its diameter, approximately equal to `3.1415926535897932`.

- `infinity` & `-infinity`

  - : An infinite value, used to indicate the largest/smallest possible value.

- `NaN`

  - : A value representing "Not a Number" canonical casing.

### Formal syntax

{{CSSSyntax}}

## Description

Mathematical constants can only be used inside [CSS math functions](/en-US/docs/Web/CSS/CSS_Functions#math_functions) for calculations. Math constants are not CSS keywords, but if they are used outside of a calculation, they're treated like any other keyword.

For example:

- `animation-name: pi;` refers to an animation named "pi", not the `pi` numeric constant.
- `line-height: e;` is invalid, but `line-height: calc(e);` is valid.

> These constants can't be used in any CSS function, only in math functions:
> 
> - `rotate(1rad * pi);` won't work. You need to use `rotate(calc(1rad * pi));`

Reviewer: ```suggestion
These constants can't be used in any CSS function, only in math functions:

- `rotate(1rad * pi);` won't work. You need to use `rotate(calc(1rad * pi));`
```

---
title: <calc-constant>
slug: Web/CSS/calc-constant
tags:
  - CSS
  - CSS Data Type
  - Data Type
  - Layout
  - Reference
  - Web
browser-compat: css.types.calc-constant
---

{{CSSRef}}

> The **`<calc-constant>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/CSS/CSS_Types) represents a well-known constants, such as `e` and `π`.

Reviewer 2: ```suggestion
The **`<calc-constant>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/CSS/CSS_Types) represents well-known constants such as `e` and `π`.
```

---
title: <calc-constant>
slug: Web/CSS/calc-constant

> tags:

Reviewer: ```suggestion
page-type: css-type
tags:
```

Rather than require authors to manually type out several digits of these constants, a few of them are provided directly by CSS.

- `e` is the base of the natural logarithm, approximately equal to `2.7182818284590452354`.
- `pi` is the ratio of a circle's circumference to its diameter, approximately equal to `3.1415926535897932`.

> **Note:** These keywords are only usable within a calculation, such as `calc(pow(e, pi) - pi)`, or `min(pi, 5, e)`. If used outside of a calculation, they're treated like any other keyword: `animation-name: pi`; refers to an animation named "pi"; `line-height: e`; is invalid (not similar to `line-height: 2.7`, but `line-height: calc(e);` is).

When a calculation or a subtree of a calculation becomes infinite or NaN, representing it with a numeric value is no longer possible. To aid in serialization of these degenerate values, the additional math constants `infinity` (with the value `+∞`), `-infinity` (with the value `−∞`), and `NaN` (with the value `NaN`) are defined.

As usual for CSS keywords, these are ASCII case-insensitive. Thus, `calc(InFiNiTy)` is perfectly valid. However, `NaN` must be serialized with this canonical casing.

> > **Note:** While not technically numbers, these keywords act as numeric values, similar to `e` and `pi`. Thus to get an infinite length, for example, requires an expression like `calc(infinity * 1px)`.

Reviewer: This seems to be a direct quote from the spec: https://w3c.github.io/csswg-drafts/css-values-4/#calc-constants. We don't generally do that, we're not adding any value there.

Things we might want to express:
- this type defines some mathematical constants for use in CSS math functions: https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Functions#math_functions
- they can only be used in these math functions
- they are case-insensitive (except `NaN`? I didn't really understand this bit)

Author: Good feedback. I've updated the text and moved it to a separate section.

## Syntax

> The syntax of `<calc-constant>` extends the syntax of {{CSSxRef("&lt;calc-value&gt;")}}.

Reviewer: I'm not sure what this is saying. "extends" would imply that valid `<calc-value>` values are valid `<calc-constant>` values, but they are not, are they?

Can you instead say something like: "The `<calc-constant` type defines a number of numeric constants that can be used in the `calc()` functional notation"?

---
title: <calc-constant>
slug: Web/CSS/calc-constant
page-type: css-type
tags:
  - CSS
  - CSS Data Type
  - Data Type
  - Layout
  - Reference
  - Web
browser-compat: css.types.calc-constant
---

{{CSSRef}}

> The **`<calc-constant>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/CSS/CSS_Types) represents well-known constants such as `e` and `π`.

Reviewer 3: ```suggestion
The **`<calc-constant>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/CSS/CSS_Types) represents well-defined constants such as `e` and `π`.
```

"well-known" may make people feel bad if they don't know them.

Rather than require authors to manually type out several digits of these constants, a few of them are provided directly by CSS.

> **Note:** These keywords are only usable within a calculation, such as `calc(pow(e, pi) - pi)`, or `min(pi, 5, e)`. If used outside of a calculation, they're treated like any other keyword: `animation-name: pi`; refers to an animation named "pi"; `line-height: e`; is invalid (not similar to `line-height: 2.7`, but `line-height: calc(e);` is).

When a calculation or a subtree of a calculation becomes infinite or NaN, representing it with a numeric value is no longer possible. To aid in serialization of these degenerate values, the additional math constants `infinity` (with the value `+∞`), `-infinity` (with the value `−∞`), and `NaN` (with the value `NaN`) are defined.

> As usual for CSS keywords, these are ASCII case-insensitive. Thus, `calc(InFiNiTy)` is perfectly valid. However, `NaN` must be serialized with this canonical casing.

Reviewer 3: ```suggestion
As usual for CSS keywords, most CSS constants are ASCII case-insensitive. Thus, `calc(InFiNiTy)` is perfectly valid. There is an exception: `NaN` must be written exactly as `NaN`, with the "Not a Number" canonical casing.
```

This needs to be easier to read than the spec ;)

> **Note:** While not technically numbers, these keywords act as numeric values, similar to `e` and `pi`. Thus to get an infinite length, for example, requires an expression like `calc(infinity * 1px)`.

## Syntax

The `<calc-constant>` type defines a number of numeric constants that can be used in CSS math functions.

### Values

- `e`
  - : The base of the natural logarithm, approximately equal to `2.7182818284590452354`.

- `pi`
  - : The ratio of a circle's circumference to its diameter, approximately equal to `3.1415926535897932`.

- `infinity` & `-infinity`
  - : An infinite length, used to indicate the largest/smallest possible value.

> - `Nan`

Reviewer 3: ```suggestion
- `NaN`
```

>   - : A value representing Not-A-Number.

Reviewer 3: ```suggestion
  - : A value representing Not-a-Number.
```

---
title: <calc-constant>
slug: Web/CSS/calc-constant
page-type: css-type
tags:
  - CSS
  - CSS Data Type
  - Data Type
  - Layout
  - Reference
  - Web
browser-compat: css.types.calc-constant
---

{{CSSRef}}

The **`<calc-constant>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/CSS/CSS_Types) represents well-defined constants such as `e` and `π`. Rather than require authors to manually type out several digits of these constants, a few of them are provided directly by CSS.

## Syntax

> The `<calc-constant>` type defines a number of numeric constants that can be used in CSS math functions.

Reviewer: ```suggestion
The `<calc-constant>` type defines a number of numeric constants that can be used in [CSS math functions](/en-US/docs/Web/CSS/CSS_Functions#math_functions).
```

### Values

- `e`
  - : The base of the natural logarithm, approximately equal to `2.7182818284590452354`.

- `pi`
  - : The ratio of a circle's circumference to its diameter, approximately equal to `3.1415926535897932`.

- `infinity` & `-infinity`

>   - : An infinite length, used to indicate the largest/smallest possible value.

Reviewer: ```suggestion
  - : An infinite value, used to indicate the largest/smallest possible value.
```

(not a necessarily a length?)

- `NaN`
  - : A value representing "Not a Number" canonical casing.

### Formal syntax

{{CSSSyntax}}

## Usage clarifications

### Used only in calculations

Mathematical constants can be used only inside [CSS math functions](/en-US/docs/Web/CSS/CSS_Functions#math_functions). If used outside of a calculation, they're treated like any other keyword.

For example, `animation-name: pi;` refers to an animation named "pi". `line-height: e;` is invalid, but `line-height: calc(e);` is valid.

### CSS data types

> While not technically numbers, css constants act as numeric values, similar to `e` and `pi`. Thus to get an infinite length, for example, requires an expression like `calc(infinity * 1px)`.

Reviewer: ```suggestion
While not technically numbers, CSS constants act as {cssxref("number")}} values. Thus to get an infinite {{cssxref("length")}}, for example, requires an expression like `calc(infinity * 1px)`.
```

"similar to `e` and `pi`" is odd - aren't `e` and `pi` just values of this? So they are not "similar to" it, they are it. Or am I confused?

> ### Case sensitivity

Author: Not sure about this request. Each topic has a separate heading because it talk about a different aspect. From a users experience perspective, visitors can skip different topics. I prefer to leave the headings.

> ### Case sensitivity

Reviewer: It's not at all necessary for every aspect in a description to have a heading, especially when the parts are as short as this. In this case:

- "Used only in calculations"  is not a good heading. Headings should describe the content below them - they are meta-content. They don't describe the thing being documented. Like say "Syntax" - it doesn't tell you what the syntax is, it tells you where to find the syntax. If you can't find a heading that describes the content, just have the content
- "CSS data types" is also not a good heading. It's not obvious to me how that heading connects up with the content at all.
- "Case sensitivity" is a fine heading

So I'd be OK to keep that last one but the others are not worth keeping IMO.

> ### Case sensitivity

Author: > * "CSS data types" is also not a good heading. It's not obvious to me how that heading connects up with the content at all.

How about "Constant data types"?

> ### Case sensitivity

Reviewer: Sorry, I don't really think this is great either. AIUI the content here is saying essentially: `<calc-constant>` values act like `<number>` values. "Constant data types" doesn't tell me that this is what that section is going to tell me.

Please, just remove the H3 heading here.

> ### Case sensitivity

Author: You provided a lot of very good modification requests for this PR, and I implemented all of them. But I disagree with you on this specific issue.

Since you said that short parts don't require headings, I enlarged the details included in each sections . We can debate what should be the heading of each part, but in my opinion, we shouldn't remove the headings.

> ### Case sensitivity

Reviewer: @ramiy , it's not our style always to have a heading for every separate topic inside the description. Chosen more or less at random:
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice#description
- https://developer.mozilla.org/en-US/docs/Web/javascript/reference/operators/import#description
- https://developer.mozilla.org/en-US/docs/Web/javascript/reference/operators/exponentiation#description
- https://developer.mozilla.org/en-US/docs/Web/CSS/@import#description
- https://developer.mozilla.org/en-US/docs/Web/CSS/animation#description
- https://developer.mozilla.org/en-US/docs/Web/CSS/@namespace#description

(I'm sure there are cases where we do have very short H3 sections in a Description, but it is not the most common practice on MDN.)

> Since you said that short parts don't require headings, I enlarged the details included in each sections .

Adding more content to justify having H3 headings is just backwards. We should write content at the level that's right for our readers, and structure it in the way that's appropriate.

I understand that you don't agree with my feedback.

>  I implemented all of them. 

Actually, you haven't. I asked for the "Examples" to be fixed and you pushed back on that, citing other pages that have the same error. I'm happy to fix those in a different PR, although I would have preferred for it to be fixed here.

> ### Case sensitivity

Reviewer 2: Hi @ramiy @wbamberg do you have any objection if I take a shot at this section? I can open a PR into Ramiy's branch (`ramiy:patch-5`), maybe it helps get this one closed out

> ### Case sensitivity

Reviewer 2: @ramiy please see my revision here: https://github.com/ramiy/content/pull/1

> ### Case sensitivity

Author: @bsmth Thank you for taking the time to review the PR and suggest new changes.

I reviewed your PR. It looks like you moved the content from the "Description" area to the intro area, removing all the H3 headings. How is that different from what @wbamberg suggested?

We disagree on one issue, **whether or not to display H3 headings**.

In my opinion, without the H3 titles we mix different topics together. At least with the titles, the user can understand what he should be considering when using those constants.

We have 3 simple topics we want to emphasize:

- **Where to use constants ?** only inside math functions.
- **What is their data type ?** two numeric constants and two degenerate numeric constants - quote from the spec.
- **Whether they are case sensitive ?** only NaN is case-sensitive.

> ### Case sensitivity

Author: Now that I'm thinking about it, we can move the entire "**Description > CSS data types**" section to "**Syntax > Return value**" to match the other math functions.

It basically describes the data type returned by the different constants.

> ### Case sensitivity

Reviewer: It's not a function though, so it doesn't have a return value.

> ### Case sensitivity

Author: Sounds reasonable, commit reverted.

> ### Case sensitivity

Reviewer 2: Hi @ramiy, I agree with @wbamberg here on the H3 headings. I will make some changes directly to this branch to help get this PR merged.

> CSS constants are ASCII case-insensitive. Thus, `calc(InFiNiTy)` is perfectly valid.

Reviewer: Why _ASCII_ case insensitive? Why not just case-insensitive?

> There is only one exception, `NaN` is case-sensitive and it must be written exactly as `NaN`. Both `nan` and `NAN` are not valid constants.

Reviewer: ```suggestion
There is only one exception: `NaN` is case-sensitive and must be written exactly as `NaN`. Both `nan` and `NAN` are invalid constants.
```

## Examples

### Valid values

```plain example-good
e
-e
E
pi
-pi
infinity
-infinity
InFiNiTy
NaN
```

### Invalid values

```plain example-bad
nan
NAN

> ```

Reviewer: I'm sorry not to have picked up on this before so I won't insist on it, but: usually I think "Examples" is to show real usage examples of the thing, not syntax illustrations. These sections would do better at the end of the `## Description`, and it would be better to have an example here where we are using these values in some CSS.

Author: https://developer.mozilla.org/en-US/docs/Web/CSS/number#examples
https://developer.mozilla.org/en-US/docs/Web/CSS/integer#examples
https://developer.mozilla.org/en-US/docs/Web/CSS/string#examples
https://developer.mozilla.org/en-US/docs/Web/CSS/image#examples

I prefer this PR to focus on `<calc-constant>`. If you want to change this, let's do it in a separate PR dedicated on unifying this across all the data-type pages.

Author: I moved the "Valid values" / "Invalid values" code blocks from the "Examples" section to "Case sensitivity" section.

- `NaN`
  - : A value representing "Not a Number" canonical casing.

### Formal syntax

{{CSSSyntax}}

## Description

### Used only in calculations

Mathematical constants can be used only inside [CSS math functions](/en-US/docs/Web/CSS/CSS_Functions#math_functions). If used outside of a calculation, they're treated like any other keyword.

For example, `animation-name: pi;` refers to an animation named "pi". `line-height: e;` is invalid, but `line-height: calc(e);` is valid.

### Constant data types

Both `e` and `pi` are numeric constants act as {{cssxref("number")}} values.

On the other end, `infinity` and `NaN` are degenerate numeric constants. While not technically numbers, they act as {{cssxref("number")}} values. Thus to get an infinite {{cssxref("length")}}, for example, requires an expression like `calc(infinity * 1px)`.

Infinite/NaN defined mostly to make serialization simpler and more obvious, but can be used to indicate a "largest possible value", since an infinite value gets clamped to the allowed range. It's rare for this to be reasonable, but when using infinity its much simpler than just putting an enormous number in a stylesheet.

### Case sensitivity

> CSS constants are case-insensitive. Thus, `calc(infinity)`, `calc(INFINITY)` and `calc(InFiNiTy)` are all perfectly valid. It applies on `e` and `pi too.

Reviewer: > It applies on `e` and `pi too.

I don't think you should say this. You already said "CSS constants are case-insensitive", and that obviously applies to all of them. If you mention `e` and `pi` here, then another value is added, we have to remember to update this list, and if we forget, people will get confused.

If you feel it's important to do something like this, try:

```suggestion
CSS constants are case-insensitive. Thus, `calc(pi)`, `calc(E)` and `calc(InFiNiTy)` are all valid.
```

...which I think makes it a bit clearer that these are supposed to be illustrative not exhaustive.
(you also don't need "perfectly").

- `animation-name: pi;` refers to an animation named "pi", not the `pi` numeric constant.
- `line-height: e;` is invalid, but `line-height: calc(e);` is valid.

> - `rotate(1rad * pi);` won't work, {{CSSxRef("transform-function/rotate", "rotate()")}} is not a math function, use `rotate(calc(1rad * pi));`

Reviewer: ```suggestion
- `rotate(1rad * pi);` won't work because {{CSSxRef("transform-function/rotate", "rotate()")}} is not a math function. Use `rotate(calc(1rad * pi));`
```
